go/merkletree2/proof_verifier.go

```diff
@@ -0,0 +1,99 @@
+package merkletree2
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+)
+
+type MerkleProofVerifier struct {
+	TreeConfig
+}
+
+func NewMerkleProofVerifier(t TreeConfig) MerkleProofVerifier {
+	return MerkleProofVerifier{TreeConfig: t}
+}
+
+func (m *MerkleProofVerifier) VerifyInclusionProof(ctx context.Context, kvp KeyValuePair, proof MerkleInclusionProof, expRootHash Hash) error {
+
+	// Hash the key value pair
+	kvpHash, err := m.hasher.HashKeyValuePairWithKeySpecificSecret(kvp, proof.KeySpecificSecret)
+	if err != nil {
+		return NewProofVerificationFailedErrorFromCause(err)
+	}
+
+	if len(proof.OtherPairsInLeaf)+1 > m.maxValuesPerLeaf {
+		return NewProofVerificationFailedError(fmt.Sprintf("Too many keys in leaf: %v", len(proof.OtherPairsInLeaf)+1))
+	}
+
+	// Reconstruct the leaf node
+	var leaf Node
+	leaf.Type = NodeTypeLeaf
+	leaf.LeafHashes = make([]KeyHashPair, len(proof.OtherPairsInLeaf)+1)
+	// LeafHashes is obtained by adding kvp into OtherPairsInLeaf while maintaining sorted order
+	for i, j, k := 0, 0, 0; i < len(proof.OtherPairsInLeaf)+1; i++ {
```

Comment:
wouldn't this be easier if the proof included what index the KV pair should go into? or even included the entire list including the kvpair instead of "Other"s?

Reply:
it is kind of confusing now with all these indices (though `k` is being used as a boolean it seems)

Reply:
It could, but we need to check the order anyways to prevent the server from putting two different values for the same key at different indexes in the tree. So it wouldn't save much...

Reply:
but it'd just be a simple linear pass